src/main/java/com/spotify/folsom/client/DefaultRawMemcacheClient.java

```diff
@@ -178,55 +176,49 @@ private DefaultRawMemcacheClient(final HostAndPort address,
 
     GLOBAL_CONNECTION_COUNT.incrementAndGet();
 
-    metrics.registerOutstandingRequestsGauge(new Metrics.OutstandingRequestsGauge() {
-      @Override
-      public int getOutstandingRequests() {
-        return pendingCounter.get();
-      }
-    });
+    metrics.registerOutstandingRequestsGauge(pendingCounter::get);
 
     channel.pipeline().addLast("handler", new ConnectionHandler());
   }
 
   @Override
-  public <T> ListenableFuture<T> send(final Request<T> request) {
+  public <T> CompletionStage<T> send(final Request<T> request) {
     if (!tryIncrementPending()) {
 
       // Do the disconnect check in here instead of outside
       // to get better performance in the happy case.
       String disconnectReason = this.disconnectReason.get();
       if (disconnectReason != null) {
         MemcacheClosedException exception = new MemcacheClosedException(disconnectReason);
-        return onExecutor(Futures.<T>immediateFailedFuture(exception));
+        return onExecutor(CompletableFutures.exceptionallyCompletedFuture(exception));
       }
 
-      return onExecutor(Futures.<T>immediateFailedFuture(new MemcacheOverloadedException(
-              "too many outstanding requests")));
+      return onExecutor(
+          CompletableFutures.exceptionallyCompletedFuture(
+              new MemcacheOverloadedException("too many outstanding requests")));
     }
     if (request instanceof SetRequest) {
       SetRequest setRequest = (SetRequest) request;
       byte[] value = setRequest.getValue();
       if (value.length > maxSetLength) {
-        return (ListenableFuture<T>) onExecutor(
-                Futures.immediateFuture(MemcacheStatus.VALUE_TOO_LARGE));
+        return (CompletionStage<T>) onExecutor(
+                CompletableFuture.completedFuture(MemcacheStatus.VALUE_TOO_LARGE));
       }
     }
     channel.write(request, new RequestWritePromise(channel, request));
     flusher.flush();
     return onExecutor(request);
   }
 
-  private <T> ListenableFuture<T> onExecutor(ListenableFuture<T> future) {
+  private <T> CompletionStage<T> onExecutor(CompletionStage<T> future) {
     return onExecutor(future, executor);
   }
 
-  private static <T> ListenableFuture<T> onExecutor(ListenableFuture<T> future, Executor executor) {
+  private static <T> CompletionStage<T> onExecutor(CompletionStage<T> future, Executor executor) {
     if (executor == null) {
       return future;
     }
-    CallbackSettableFuture<T> newFuture = new CallbackSettableFuture<>(future);
-    future.addListener(newFuture, executor);
-    return newFuture;
+    return future.thenApplyAsync(Function.identity(), executor);
```

Comment:
I am not sure if this is correct - will exceptions run on this executor? How could I make that happen?

Reply:
Docs and a quick empirical test indicates that the executor supplied to `thenApplyAsync` is not used for exceptions.

The `handleAsync` method seems to cause both normal values and exceptions to be run on the supplied executor.

E.g.

```java
return future.handleAsync((v, e) -> {
  if (v != null) {
    return v;
  } else {
    throw new CompletionException(e);
  }
}
```

Reply:
Even better:

```java
return future.whenCompleteAsync((v, e) -> { }, executor);
```